### DISCUSSION TITLE
Having issues running the host tool

### DISCUSSION THREAD

**Discussion by ShaharHD** [rauc/rauc#1148] [Q&A]
When adding the RAUC meta to a PetaLinux (Xilinx Yocto) on a Ubuntu 20.04 it builds correctly

```shell
petalinux-build -c rauc-native
```

But when executing it fails with

```shell
./rauc: symbol lookup error: ./rauc: undefined symbol: g_log_writer_default_set_use_stderr
```

Current versions installed
```shell
libglib2.0-0/focal-updates,now 2.64.6-1~ubuntu20.04.4 amd64 [installed]
  GLib library of C routines

libglib2.0-dev/focal-updates,now 2.64.6-1~ubuntu20.04.4 amd64 [installed]
  Development files for the GLib library

libglib2.0-dev-bin/focal-updates,now 2.64.6-1~ubuntu20.04.4 amd64 [installed,automatic]
  Development utilities for the GLib library
```

Seems to be this one: https://github.com/rauc/rauc/blob/43c1e11152b4e36bb608f3cc4de9d4cfa397f8bd/src/main.c#L2528

Looking more into this, the PetaLinux/Yocto has a newer version 
(http://layers.openembedded.org/layerindex/recipe/190013/) but this one is only for the `TARGET` (and not the `HOST`) AFAIK

Any idea as to why it fails?

We are using PetaLinux 2022.2 which supports:
```
Ubuntu Linux Workstation/Server 18.04.1, 18.04.2, 18.04.3, 18.04.4, 18.04.5, 18.04.06, 20.04, 20.04.1, 20.04.2, 20.04.3 ,20.04.4 (64-bit)
```

So I would like to refrain from re-installing the the whole system with 22.04 ...

**Comment by ejoerns** [comment]
I would assume that you are running RAUC native binary (which is built against the Petalinux libraries) against your host (ubuntu 22.04) libraries. This may or may not work. In this case it doesn't because the symbol `g_log_writer_default_set_use_stderr()` is available in glib 2.68, but ubuntu 22.04 has glib [2.64](https://packages.ubuntu.com/focal/libglib2.0-0).

The solution to your problem would be to run rauc-native against the native dependencies from your Petalinux (i.e. Yocto) build. With plain Yocto, this would work with the `oe-run-native` helper script as described here:

https://github.com/rauc/meta-rauc#ii-building-and-using-rauc-host-tool

Not sure how this maps to Petalinux. But petalinux just wraps bitbake calls, so it might be possible.

**Reply by ShaharHD** [reply]
I am using https://github.com/rauc/meta-rauc/tree/honister as a layer in my Petalinux build (matched to PetaLinux 2022.2 which is based on Honister)

And I was following the exact step you were referrning to (https://github.com/rauc/meta-rauc/tree/honister#ii-building-rauc-host-tool),
Petalinux wraps around Yocto, so it does call bitbake in the background when calling 

```shell
petalinux-build -c rauc-native
```

It seems when petalinux builds the native application, it does generates the correct host executable, but for some reason it uses the h files of the Target system in the build and not the host (which is why it thinks its `2.68.4` and not `2.64.6`) - maybe something with incldues search order?

If you wish to reproduce this, you can download petalinux from xilinx (free registreation), install it on a ubuntu 20.04.xx sysetm and I can supply you with a small git repo which you can pull and try to build.

I'll also try and go the manual bitbake route of PetaLinux and see if it makes any difference:
https://docs.xilinx.com/r/2022.2-English/ug1144-petalinux-tools-reference-guide/Accessing-BitBake/Devtool-in-a-Project

**Reply by ejoerns** [reply]
For native builds, bitbake builds against its native sysroot which contains the library version provided by the corresponding recipes.

For `honister`, it builds against the glib-native-dev provided by honister. This is version [2.68.4](http://layers.openembedded.org/layerindex/recipe/190013/)

So I do not see any issue here. As I noted above, the reason it does not run when being simply executed because this would use the host system's sysroot and *not* the native bitbake sysroot. [oe-run-native](https://git.yoctoproject.org/poky/plain/scripts/oe-run-native) instead uses the native sysroot.

**Reply by ejoerns** [reply]
Also note, when using `rauc` from petalinux, then probably for generating bundles with `bundle.bbclass`. This should work anyway.

**Reply by ShaharHD** [reply]
I'm a bit lost here ...  but managed to make it work based on your inputs but now I have new insights.

Followed your advice and I can now see where `libglib-2.0.so.0` for the native is stored at when I export the SDK (not easily available when using the yocto build folder)

So if i'm running it using the `LD_LIBRARY_PATH` is works :)

```shell
LD_LIBRARY_PATH=<extracted sdk path>/sysroots/x86_64-petalinux-linux/usr/lib ./build/tmp/deploy/tools/rauc
```
Which is why it will probably work when running via the automated yocto builds, but I do require an easy access to the native `rauc` command.

Maybe the simplest solution for the local native tool is to simply clone the rauc repo and build it locally?

1. Can I take the latest version 1.9 or I have to stick to 1.7 (as set by honister on `meta-rauc`)?
2. Are there any specific settings or build flags you pass when you build via the `meta-rauc`?

FFR who ever reads this, I followed https://docs.xilinx.com/r/2022.2-English/ug1144-petalinux-tools-reference-guide/Accessing-BitBake/Devtool-in-a-Project and verfied the resulting file from `bitbake native-rauc` and `petalinux-build -c rauc-native` is identical)

**Reply by ejoerns** [reply]
> I'm a bit lost here ... but managed to make it work based on your inputs but now I have new insights.
> 
> Followed your advice and I can now see where `libglib-2.0.so.0` for the native is stored at when I export the SDK (not easily available when using the yocto build folder)
> 
> So if i'm running it using the `LD_LIBRARY_PATH` is works :)
> 
> ```shell
> LD_LIBRARY_PATH=<extracted sdk path>/sysroots/x86_64-petalinux-linux/usr/lib ./build/tmp/deploy/tools/rauc
> ```

Yes, this sounds like the proper manual way of enforcing the library search path.
It should also be possible to do without using the SDK when running the equivalent of

    bitbake rauc-native -c addto_recipe_sysroot

before.

> Which is why it will probably work when running via the automated yocto builds, but I do require an easy access to the native `rauc` command.
> 
> Maybe the simplest solution for the local native tool is to simply clone the rauc repo and build it locally?

That would be an alternative, yes.

> 1. Can I take the latest version 1.9 or I have to stick to 1.7 (as set by honister on `meta-rauc`)?

You should be able to take the latest version. They are designed to be compatible with regard to the created bundles:

https://rauc.readthedocs.io/en/latest/faq.html#is-the-rauc-bundle-format-forwards-backwards-compatible

> 2. Are there any specific settings or build flags you pass when you build via the `meta-rauc`?

For the native build you could disable some non-required options that would just pull in more dependencies. But you can also use the defaults.

What should work is:

    meson setup -Dservice=false -Dnetwork=false -Dgpt=disabled -Dstreaming=false build

> FFR who ever reads this, I followed https://docs.xilinx.com/r/2022.2-English/ug1144-petalinux-tools-reference-guide/Accessing-BitBake/Devtool-in-a-Project and verfied the resulting file from `bitbake native-rauc` and `petalinux-build -c rauc-native` is identical)
